<a href="https://colab.research.google.com/github/cmburban/MAT-422/blob/main/MAT422_Section_1_2_Linear_Algebra.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MAT 422 - Section 1.2: Linear Algebra Concepts in Python

Topics: linear spaces, orthogonality, Gram-Schmidt process, eigenvalues and eigenvectors

## 1.2.1 Linear Spaces

Linear spaces are sets of vectors that are closed under vector addition and scalar multiplication. In this section, I will demonstrate linear combinations, span, and linear independence using NumPy.

In [6]:
import numpy as np

In [12]:
u = np.array([1, 2])
v = np.array([3, 1])

print("u =", u)
print("v =", v)

u = [1 2]
v = [3 1]


### Linear Combination

Linear combinations are formed by multiplying vectors by scalar values, then adding the results. I will calculate 2u + 3v as an example.

In [13]:
linear_combination = 2 * u + 3 * v

print("2u + 3v =", linear_combination)

2u + 3v = [11  7]


The result is [11, 7] showing that a new vector can be formed by multiplying the original vectors by scalar values and adding them together.

### Span

The span of a set of vectors is the set of all vectors that can be created from their linear combinations. In this example, I will use different scalar values to create several vectors from u and v.

In [14]:
combinations = [
    1 * u + 1 * v,
    2 * u + 1 * v,
    1 * u + 2 * v,
    -1 * u + 2 * v
]

for i, result in enumerate(combinations, 1):
    print(f"Combination {i}: {result}")

Combination 1: [4 3]
Combination 2: [5 5]
Combination 3: [7 4]
Combination 4: [5 0]


Each result above is a different linear combination of u and v. Since the scalar values can be changed, many different vectors can be made. The collection of all vectors that can be made this way is the span of u and v.

### Linear Independence

A set of vectors is linearly independent if none of the vectors can be written as a linear combination of the others. For vectors u and v, I will check whether one vector is a scalar multiple of the other.

In [15]:
# Check whether v is a scalar multiple of u
ratio = v[0] / u[0]

print("Ratio between corresponding first components:", ratio)
print("Ratio between corresponding second components:", v[1] / u[1])

if np.isclose(v[1] / u[1], ratio):
    print("u and v are linearly dependent.")
else:
    print("u and v are linearly independent.")

Ratio between corresponding first components: 3.0
Ratio between corresponding second components: 0.5
u and v are linearly independent.


The ratios of the corresponding components are different, so v cannot be written as a scalar multiple of u. This means u and v are linearly independent. This means that neither vector can be produced by scaling the other.

### Basis and Dimension

Because u and v are linearly independent and are vectors in R², they form a basis for R². So, the dimension of the space they span is 2.

In [17]:
A = np.column_stack((u, v))

print("Matrix formed from u and v:")
print(A)
print("Rank:", np.linalg.matrix_rank(A))

Matrix formed from u and v:
[[1 3]
 [2 1]]
Rank: 2


## 1.2.2 Orthogonality

Orthogonality is when two vectors are perpendicular. Two vectors are orthogonal when their dot product is zero. I will show the dot product, vector norms, and projection using NumPy.

In [18]:
a = np.array([1, 2])
b = np.array([2, -1])

dot_product = np.dot(a, b)

print("a =", a)
print("b =", b)
print("Dot product:", dot_product)

a = [1 2]
b = [ 2 -1]
Dot product: 0


The dot product of a and b is 0. The two vectors are orthogonal, which means they are perpendicular to each other.

### Vector Norms

The norm of a vector shows its length. For a vector x, the Euclidean norm is shown as

$$
\|x\| = \sqrt{x_1^2 + x_2^2 + \cdots + x_n^2}
$$

I will find the norms of a and b.

In [19]:
norm_a = np.linalg.norm(a)
norm_b = np.linalg.norm(b)

print("||a|| =", norm_a)
print("||b|| =", norm_b)

||a|| = 2.23606797749979
||b|| = 2.23606797749979


Both vectors have the same norm, about 2.236. The norm measures the length or magnitude of a vector.

### Projection

The projection of one vector onto another gives the component of one vector that lies in the direction of the other. The projection of a onto b is

$$
\operatorname{proj}_b(a) =
\frac{a \cdot b}{b \cdot b}b
$$

I will calculate this projection using NumPy.

In [20]:
projection = (np.dot(a, b) / np.dot(b, b)) * b

print("Projection of a onto b:", projection)

Projection of a onto b: [ 0. -0.]


The projection is the zero vector because a and b are orthogonal. Since a has no component in the direction of b, its projection onto b is 0.

### Checking for Orthogonality

Two vectors are orthogonal when their dot product is 0. I will check the dot product above.

In [21]:
if np.isclose(np.dot(a, b), 0):
    print("The vectors are orthogonal.")
else:
    print("The vectors are not orthogonal.")

The vectors are orthogonal.


The dot product is 0, confirming that a and b are orthogonal. The projection of a onto b is also the zero vector, which is consistent with the fact that a has no component in the direction of b.

## 1.2.3 Gram-Schmidt Process

The Gram-Schmidt Process converts a set of linearly independent vectors to an orthonormal set of vectors that spans the same space. I will apply this process to three vectors in R3.

In [22]:
v1 = np.array([1, 1, 0], dtype=float)
v2 = np.array([1, 0, 1], dtype=float)
v3 = np.array([0, 1, 1], dtype=float)

print("v1 =", v1)
print("v2 =", v2)
print("v3 =", v3)

v1 = [1. 1. 0.]
v2 = [1. 0. 1.]
v3 = [0. 1. 1.]


### Step 1: First Orthogonal Vector

The first orthogonal vector is the first vector. I will call it u1.

In [23]:
u1 = v1

print("u1 =", u1)

u1 = [1. 1. 0.]


### Step 2: Second Orthogonal Vector

To make the second vector orthogonal to the first, I subtract the projection of v2 onto u1 from v2.

In [24]:
projection_v2 = (np.dot(v2, u1) / np.dot(u1, u1)) * u1
u2 = v2 - projection_v2

print("Projection of v2 onto u1:", projection_v2)
print("u2 =", u2)

Projection of v2 onto u1: [0.5 0.5 0. ]
u2 = [ 0.5 -0.5  1. ]


### Step 3: Third Orthogonal Vector

For the third vector, I subtract its projections onto both u1 and u2. This makes the resulting vector orthogonal to both previous vectors.

In [25]:
projection_v3_u1 = (np.dot(v3, u1) / np.dot(u1, u1)) * u1
projection_v3_u2 = (np.dot(v3, u2) / np.dot(u2, u2)) * u2

u3 = v3 - projection_v3_u1 - projection_v3_u2

print("Projection onto u1:", projection_v3_u1)
print("Projection onto u2:", projection_v3_u2)
print("u3 =", u3)

Projection onto u1: [0.5 0.5 0. ]
Projection onto u2: [ 0.16666667 -0.16666667  0.33333333]
u3 = [-0.66666667  0.66666667  0.66666667]


### Step 4: Normalize the Orthogonal Vectors

To create an orthonormal set, I must divide each orthogonal vector by its norm. This gives each vector a length of 1 while keeping its direction.

In [26]:
q1 = u1 / np.linalg.norm(u1)
q2 = u2 / np.linalg.norm(u2)
q3 = u3 / np.linalg.norm(u3)

print("q1 =", q1)
print("q2 =", q2)
print("q3 =", q3)

q1 = [0.70710678 0.70710678 0.        ]
q2 = [ 0.40824829 -0.40824829  0.81649658]
q3 = [-0.57735027  0.57735027  0.57735027]


### Checking the Orthonormal Basis

The vectors should each have a norm of 1 and should be orthogonal. I will check to make sure this is true.

In [27]:
print("Norms:")
print(np.linalg.norm(q1))
print(np.linalg.norm(q2))
print(np.linalg.norm(q3))

print("\nDot products:")
print("q1 · q2 =", np.dot(q1, q2))
print("q1 · q3 =", np.dot(q1, q3))
print("q2 · q3 =", np.dot(q2, q3))

Norms:
0.9999999999999999
1.0
0.9999999999999998

Dot products:
q1 · q2 = 0.0
q1 · q3 = 0.0
q2 · q3 = 1.1102230246251565e-16


The three vectors all have a norm of 1, and their dot products are 0. Therefore, the Gram-Schmidt Process did make an orthonormal set of vectors.

In [32]:
Q = np.column_stack((q1, q2, q3))

print("Q =")
print(Q)

print("\nQ.T @ Q =")
print(Q.T @ Q)

Q =
[[ 0.70710678  0.40824829 -0.57735027]
 [ 0.70710678 -0.40824829  0.57735027]
 [ 0.          0.81649658  0.57735027]]

Q.T @ Q =
[[1.00000000e+00 2.45142679e-17 2.73204007e-17]
 [2.45142679e-17 1.00000000e+00 9.80784505e-17]
 [2.73204007e-17 9.80784505e-17 1.00000000e+00]]


The matrix Q has the three orthonormal vectors as columns. Since Q.T @ Q is about equal to the identity matrix, the vectors are orthonormal.

## 1.2.4 Eigenvalues and Eigenvectors

For a square matrix A, an eigenvector is a nonzero vector which direction is not changed when multiplied by A. The multiplication only scales the vector. This relationship is written as A v = lambda v, where lambda is the eigenvalue.

In [28]:
A = np.array([
    [3, 1],
    [1, 3]
], dtype=float)

print("Matrix A:")
print(A)

Matrix A:
[[3. 1.]
 [1. 3.]]


### Calculating Eigenvalues and Eigenvectors

I will use NumPy to calculate the eigenvalues and corresponding eigenvectors of A.

In [29]:
eigenvalues, eigenvectors = np.linalg.eig(A)

print("Eigenvalues:")
print(eigenvalues)

print("\nEigenvectors:")
print(eigenvectors)

Eigenvalues:
[4. 2.]

Eigenvectors:
[[ 0.70710678 -0.70710678]
 [ 0.70710678  0.70710678]]


### Checking the Eigenpairs

For each eigenvalue and eigenvector, I will check that A v = lambda v. The residual A v - lambda v should be close to zero.

In [33]:
for i, eigenvalue in enumerate(eigenvalues):
    vector = eigenvectors[:, i]

    Av = A @ vector
    lambda_v = eigenvalue * vector
    residual = Av - lambda_v

    print(f"Eigenpair {i + 1}")
    print("Eigenvalue:", eigenvalue)
    print("Eigenvector:", vector)
    print("A @ v:", Av)
    print("lambda * v:", lambda_v)
    print("Residual norm:", np.linalg.norm(residual))
    print()

Eigenpair 1
Eigenvalue: 4.0
Eigenvector: [0.70710678 0.70710678]
A @ v: [2.82842712 2.82842712]
lambda * v: [2.82842712 2.82842712]
Residual norm: 0.0

Eigenpair 2
Eigenvalue: 2.0
Eigenvector: [-0.70710678  0.70710678]
A @ v: [-1.41421356  1.41421356]
lambda * v: [-1.41421356  1.41421356]
Residual norm: 0.0



The residual norms are 0, showing that the calculated eigenvalues and eigenvectors do satisfy A v = lambda v. The eigenvectors show directions that are retained by the matrix, while the eigenvalues show how those directions are scaled.

### Orthogonality of the Eigenvectors

Since A is symmetric, its eigenvectors should be orthogonal. I will check this using their dot product.

In [31]:
v1 = eigenvectors[:, 0]
v2 = eigenvectors[:, 1]

print("Dot product of the eigenvectors:", np.dot(v1, v2))

Dot product of the eigenvectors: 0.0


The dot product of the two eigenvectors is 0, which means they are orthogonal. This is consistent with the property that eigenvectors that correspond to certain: eigenvalues of a symmetric matrix are orthogonal.

## End of Notebook Reflection

Although I had taken Linear Algebra as a course at ASU beforehand, I never understood how data science can be applied to their concepts. Starting with linear combinations, I learned how vectors can be defined by their space with span, as well as direction through linear independence. Dot products were interesting in that they showed how we can compare different vectors. Probably my most favorite part of this homework was demonstrating the Gran-Schmidt Process, in which independent vectors can be changed to an orthonormal basis. Learning about eigenvectors and eigenvalues helped me understand how scaling vectors can effect their directions. Overall, manipulating examples to demonstrate these ideas helped me learn a ton about linear algebra.

## Reference

[Linear Algebra Slides](https://canvas.asu.edu/courses/253497/pages/linear-algebra-slides?module_item_id=19107975$0)¹
 [ChatGPT (2026)](https://chatgpt.com)²

## Checklist

I explained each concept in words before or after the code. [DONE]

I verified important claims numerically, such as dot products, norms, identities, or residuals. [DONE]

I changed or extended at least one example so the notebook reflects my own work. [DONE]

I ran the notebook from top to bottom without errors. [DONE]

I saved the notebook from Google Colab directly to GitHub, following the course instructions. [DONE]